transport-native-io_uring/src/main/java/io/netty/channel/uring/AbstractIOUringChannel.java

```diff
@@ -0,0 +1,250 @@
+    public ChannelMetadata metadata() {
+        return METADATA;
+    }
+
+    @Override
+    protected abstract AbstractUringUnsafe newUnsafe();
+
+    @Override
+    protected boolean isCompatible(final EventLoop loop) {
+        return loop instanceof IOUringEventLoop;
+    }
+
+    public void doReadBytes(ByteBuf byteBuf) {
+        IOUringEventLoop ioUringEventLoop = (IOUringEventLoop) eventLoop();
+        IOUringSubmissionQueue submissionQueue = ioUringEventLoop.getRingBuffer().getIoUringSubmissionQueue();
+
+        unsafe().recvBufAllocHandle().attemptedBytesRead(byteBuf.writableBytes());
+
+        if (byteBuf.hasMemoryAddress()) {
+            long eventId = ioUringEventLoop.incrementEventIdCounter();
+            final Event event = new Event();
+            event.setId(eventId);
+            event.setOp(EventType.READ);
+            event.setReadBuffer(byteBuf);
+            event.setAbstractIOUringChannel(this);
+            submissionQueue.add(eventId, EventType.READ, socket.getFd(), byteBuf.memoryAddress(),
+                                byteBuf.writerIndex(), byteBuf.capacity());
+        }
+    }
+
+    protected final ByteBuf newDirectBuffer(ByteBuf buf) {
+        return newDirectBuffer(buf, buf);
+    }
+
+    protected final ByteBuf newDirectBuffer(Object holder, ByteBuf buf) {
+        final int readableBytes = buf.readableBytes();
+        if (readableBytes == 0) {
+            ReferenceCountUtil.release(holder);
+            return Unpooled.EMPTY_BUFFER;
+        }
+
+        final ByteBufAllocator alloc = alloc();
+        if (alloc.isDirectBufferPooled()) {
+            return newDirectBuffer0(holder, buf, alloc, readableBytes);
+        }
+
+        final ByteBuf directBuf = ByteBufUtil.threadLocalDirectBuffer();
+        if (directBuf == null) {
+            return newDirectBuffer0(holder, buf, alloc, readableBytes);
+        }
+
+        directBuf.writeBytes(buf, buf.readerIndex(), readableBytes);
+        ReferenceCountUtil.safeRelease(holder);
+        return directBuf;
+    }
+
+    private static ByteBuf newDirectBuffer0(Object holder, ByteBuf buf, ByteBufAllocator alloc, int capacity) {
+        final ByteBuf directBuf = alloc.directBuffer(capacity);
+        directBuf.writeBytes(buf, buf.readerIndex(), capacity);
+        ReferenceCountUtil.safeRelease(holder);
+        return directBuf;
+    }
+
+    @Override
+    protected void doDisconnect() throws Exception {
+    }
+
+    @Override
+    protected void doClose() throws Exception {
+    }
+
+    // Channel/ChannelHandlerContext.read() was called
+    @Override
+    protected void doBeginRead() throws Exception {
+        final AbstractUringUnsafe unsafe = (AbstractUringUnsafe) unsafe();
+        if (!uringInReadyPending) {
+            uringInReadyPending = true;
+            unsafe.executeUringReadOperator();
+        }
+    }
+
+    @Override
+    protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+        if (in.size() == 1) {
```

Comment:
@normanmaurer yeah the problem is  I cant read the next `ByteBuf` because I must first remove the bytes in ChannelOutboundBuffer
I need to map that somehow in EventLoop, I'll think about it tomorrow, to keep it simple I thought one write should be enough